internal/os/os.go

```diff
@@ -1,12 +1,260 @@
+	GreenbooFailFileName             = "/etc/greenboot/red.d/bootfail.sh"
+	StartGracefulReboot              = "StartGracefulReboot"
+	FinishedGracefulReboot           = "FinishedGracefulReboot"
+	TimeoutToGracefulRebootInSeconds = 10
+	EdgeConfFileName                 = "/etc/ostree/remotes.d/edge.conf"
 )
 
+type Deployments struct {
+	Checksum string `json:"checksum"`
+	Timestamp int `json:"timestamp"`
+	Booted bool `json:"booted"`
+}
+
+type StatusStruct struct {
+	Deployments []*Deployments `json:"deployments"`
+}
+
 type OS struct {
+	AutomaticallyUpgrade                bool
+	OsCommit                            string
+	RequestedOsCommit                   string
+	HostedImagesURL                     string
+	LastUpgradeStatus                   string
+	LastSucceededUpgradeTime            string
+	LastUpgradeTime                     string
+	GracefulRebootChannel               chan string
+	WaitTillFinishGracefulRebootChannel chan string
+}
+
+func NewOS(gracefulRebootChannel chan string) *OS {
+	waitTillFinishGracefulRebootChannel := make(chan string)
+	return &OS{
+		AutomaticallyUpgrade:                true,
+		OsCommit:                            UnknownOsImageId,
+		RequestedOsCommit:                   UnknownOsImageId,
+		HostedImagesURL:                     "",
+		GracefulRebootChannel:               gracefulRebootChannel,
+		WaitTillFinishGracefulRebootChannel: waitTillFinishGracefulRebootChannel,
+	}
+}
+
+func (o *OS) UpdateRpmOstreeStatus()  {
+	cmd := exec.Command("rpm-ostree", "status", "--json")
+	stdout, err := cmd.Output()
+
+	if err != nil {
+		log.Error("Failed to run 'rpm-ostree status'")
+	}
+
+	resUnMarsh := StatusStruct{}
+	if err := json.Unmarshal([]byte(stdout), &resUnMarsh); err != nil{
+		log.Error("Failed to unmarshal json")
+	}
+
+	if len(resUnMarsh.Deployments) == 0 {
+		log.Error("Failed to unmarshal json")
+	}
+
+	// go over the deployments, check which one is the booted one and if it's the required
+	for _, deployment := range resUnMarsh.Deployments{
+		if deployment.Booted {
+			o.OsCommit = deployment.Checksum
+			o.LastSucceededUpgradeTime = time.Unix(int64(deployment.Timestamp),0).String()
+			if o.RequestedOsCommit == UnknownOsImageId{
+				o.LastUpgradeTime = time.Unix(int64(deployment.Timestamp),0).String()
+				o.LastUpgradeStatus = "succeeded"
+			}
+		}
+
+		if deployment.Checksum == o.RequestedOsCommit{
+			o.LastUpgradeTime = time.Unix(int64(deployment.Timestamp),0).String()
+			if !deployment.Booted {
+				o.LastUpgradeStatus = "failed"
+			} else {
+				o.LastUpgradeStatus = "succeeded"
+			}
+		}
+	}
+}
+
+func (o *OS) Init(config models.DeviceConfigurationMessage) error {
+	return o.Update(config)
+}
+
+func (o *OS) Update(configuration models.DeviceConfigurationMessage) error {
+	newOSInfo := configuration.Configuration.Os
+	if newOSInfo == nil {
+		return fmt.Errorf("cannot retrieve configuration info.")
+	}
+
+	if newOSInfo.HostedImagesURL != o.HostedImagesURL {
```

Comment:
What should happen if automaticUpgrade is set to false? or to true? 
Shouldn't the `AutomaticUpdatePolicy` be updated accordingly not to/to trigger auto update by rpm-ostreed-automatic.timer? 
https://www.mankier.com/8/rpm-ostreed-automatic.service
https://www.mankier.com/5/rpm-ostreed.conf

Reply:
updated it